### ISSUE TITLE
java.io.NotSerializableException: org.activiti.engine.impl.variable.StringType

### ISSUE THREAD

**Issue by goozi** [Activiti/Activiti#900]
activiti:5.19.0
db:mysql5.7
orm:hibernate4.2.5

When i complete task, the error occur:

workFlowService.completeNextUserTask(processInstanceId,map);

map:
"userid", "1111-2222-33333333-4444"
"var1", true

1 when i put one parameter in map ，it works fine。
2 when i put one more parameters in map，occur error 。
### Cause: java.sql.SQLException: Invalid argument value: java.io.NotSerializableException

```
at org.apache.ibatis.exceptions.ExceptionFactory.wrapException(ExceptionFactory.java:23)
at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:147)
at org.apache.ibatis.session.defaults.DefaultSqlSession.insert(DefaultSqlSession.java:134)
at org.activiti.engine.impl.db.DbSqlSession.flushBulkInsert(DbSqlSession.java:843)
at org.activiti.engine.impl.db.DbSqlSession.flushPersistentObjects(DbSqlSession.java:812)
at org.activiti.engine.impl.db.DbSqlSession.flushInserts(DbSqlSession.java:790)
at org.activiti.engine.impl.db.DbSqlSession.flush(DbSqlSession.java:611)
at org.activiti.engine.impl.interceptor.CommandContext.flushSessions(CommandContext.java:212)
at org.activiti.engine.impl.interceptor.CommandContext.close(CommandContext.java:138)
at org.activiti.engine.impl.interceptor.CommandContextInterceptor.execute(CommandContextInterceptor.java:66)
at org.activiti.spring.SpringTransactionInterceptor$1.doInTransaction(SpringTransactionInterceptor.java:47)
at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
at org.activiti.spring.SpringTransactionInterceptor.execute(SpringTransactionInterceptor.java:45)
at org.activiti.engine.impl.interceptor.LogInterceptor.execute(LogInterceptor.java:37)
at org.activiti.engine.impl.cfg.CommandExecutorImpl.execute(CommandExecutorImpl.java:40)
at org.activiti.engine.impl.cfg.CommandExecutorImpl.execute(CommandExecutorImpl.java:35)
at org.activiti.engine.impl.TaskServiceImpl.complete(TaskServiceImpl.java:179)
at com.qihang.winter.web.activiti.service.impl.WorkFlowServiceImpl.completeNextUserTask(WorkFlowServiceImpl.java:109)
at sun.reflect.GeneratedMethodAccessor963.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java)
at java.lang.reflect.Method.invoke(Method.java:606)
at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:98)
at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:262)
at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:95)
at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
at com.sun.proxy.$Proxy82.completeNextUserTask(Unknown Source)
at com.qihang.winter.web.cgform.controller.build.CgFormBuildController.complet(CgFormBuildController.java:594)
at com.qihang.winter.web.cgform.controller.build.CgFormBuildController.saveOrUpdate(CgFormBuildController.java:366)
at sun.reflect.GeneratedMethodAccessor793.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java)
at java.lang.reflect.Method.invoke(Method.java:606)
at org.springframework.web.method.support.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:215)
at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:132)
at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:104)
at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandleMethod(RequestMappingHandlerAdapter.java:749)
at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:689)
at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:83)
at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:938)
at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:870)
at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:961)
at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:863)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:650)
at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:837)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
at com.qihang.winter.core.aop.GZipFilter.doFilter(GZipFilter.java:49)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
at com.alibaba.druid.support.http.WebStatFilter.doFilter(WebStatFilter.java:140)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:88)
at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
at org.springframework.orm.hibernate4.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:150)
at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
at org.apache.catalina.core.StandardContextValve.__invoke(StandardContextValve.java:122)
at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java)
at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
at org.apache.catalina.core.StandardHostValve.__invoke(StandardHostValve.java:170)
at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java)
at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:423)
at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)
at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:620)
at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:318)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
at java.lang.Thread.run(Thread.java:745)
```

Caused by: java.sql.SQLException: Invalid argument value: java.io.NotSerializableException
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:1078)
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:989)
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:975)
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:920)
    at com.mysql.jdbc.PreparedStatement.setSerializableObject(PreparedStatement.java:4413)
    at com.mysql.jdbc.PreparedStatement.setObject(PreparedStatement.java:4084)
    at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_setObject(FilterChainImpl.java:2917)
    at com.alibaba.druid.filter.FilterAdapter.preparedStatement_setObject(FilterAdapter.java:1281)
    at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_setObject(FilterChainImpl.java:2914)
    at com.alibaba.druid.proxy.jdbc.PreparedStatementProxyImpl.setObject(PreparedStatementProxyImpl.java:359)
    at com.alibaba.druid.pool.DruidPooledPreparedStatement.setObject(DruidPooledPreparedStatement.java:460)
    at sun.reflect.GeneratedMethodAccessor1095.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.apache.ibatis.logging.jdbc.PreparedStatementLogger.invoke(PreparedStatementLogger.java:66)
    at com.sun.proxy.$Proxy73.setObject(Unknown Source)
    at org.apache.ibatis.type.ObjectTypeHandler.setNonNullParameter(ObjectTypeHandler.java:28)
    at org.apache.ibatis.type.BaseTypeHandler.setParameter(BaseTypeHandler.java:46)
    at org.apache.ibatis.type.UnknownTypeHandler.setNonNullParameter(UnknownTypeHandler.java:37)
    at org.apache.ibatis.type.BaseTypeHandler.setParameter(BaseTypeHandler.java:46)
    at org.apache.ibatis.executor.parameter.DefaultParameterHandler.setParameters(DefaultParameterHandler.java:91)
    at org.apache.ibatis.executor.statement.PreparedStatementHandler.parameterize(PreparedStatementHandler.java:77)
    at org.apache.ibatis.executor.statement.RoutingStatementHandler.parameterize(RoutingStatementHandler.java:58)
    at org.apache.ibatis.executor.SimpleExecutor.prepareStatement(SimpleExecutor.java:71)
    at org.apache.ibatis.executor.SimpleExecutor.doUpdate(SimpleExecutor.java:44)
    at org.apache.ibatis.executor.BaseExecutor.update(BaseExecutor.java:108)
    at org.apache.ibatis.executor.CachingExecutor.update(CachingExecutor.java:75)
    at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:145)
    ... 82 more
Caused by: java.io.NotSerializableException: org.activiti.engine.impl.variable.StringType
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1183)
    at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347)
    at com.mysql.jdbc.PreparedStatement.setSerializableObject(PreparedStatement.java:4402)
    ... 105 more

Pls kind to help!

**Comment by jbarrez** [comment]
> workFlowService.completeNextUserTask(processInstanceId,map);
 That's not an Activiti service .. do you have a unit test using only the Activiti API that demonstrates this problem?

**Comment by goozi** [comment]
public void completeNextUserTask(String pid,Map<String,Object> variables){
        Task task = taskService.createTaskQuery().processInstanceId(pid).active().singleResult();
            taskService.complete(task.getId(), variables);
    }

**Comment by jbarrez** [comment]
sure ... but with what variables are you doing this? That's what's going wrong here.

**Comment by goozi** [comment]
Actually completeNextUserTask method is only two lines of code, the error is the same!

**Comment by jbarrez** [comment]
I mean, what kind of variables are you passing in the map? The exception

Caused by: java.io.NotSerializableException: org.activiti.engine.impl.variable.StringType

clearly says its an error related to persisting the variable.

**Comment by goozi** [comment]
<process id="test9" name="test9" isExecutable="true">
    <startEvent id="s" name="s" activiti:initiator="applyUserId"/>
    <userTask id="a" name="a" activiti:assignee="${applyUserId}"/>
    <userTask id="b" name="b" activiti:assignee="${auditor1}"/>
    <exclusiveGateway id="j1" name="j1"/>
    <userTask id="c" name="c" activiti:assignee="${auditor2}"/>
    <exclusiveGateway id="j2" name="j2"/>
    <endEvent id="e" name="e"/>
    <sequenceFlow id="f1" name="f1" sourceRef="s" targetRef="a"/>
    <sequenceFlow id="f2" name="f2" sourceRef="a" targetRef="b"/>
    <sequenceFlow id="f3" name="f3" sourceRef="b" targetRef="j1"/>
    <sequenceFlow id="f5" name="f5" sourceRef="j1" targetRef="c">
      <conditionExpression xsi:type="tFormalExpression"><![CDATA[${j1}]]></conditionExpression>
    </sequenceFlow>
    <sequenceFlow id="f4" name="f4" sourceRef="j1" targetRef="a">
      <conditionExpression xsi:type="tFormalExpression"><![CDATA[${!j1}]]></conditionExpression>
    </sequenceFlow>
    <sequenceFlow id="f6" name="f6" sourceRef="c" targetRef="j2"/>
    <sequenceFlow id="f7" name="f7" sourceRef="j2" targetRef="b">
      <conditionExpression xsi:type="tFormalExpression"><![CDATA[${!j2}]]></conditionExpression>
    </sequenceFlow>
    <sequenceFlow id="f8" name="f8" sourceRef="j2" targetRef="e">
      <conditionExpression xsi:type="tFormalExpression"><![CDATA[${j2}]]></conditionExpression>
    </sequenceFlow>
  </process>


map:
 "auditor1", "1111-2222-33333333-4444"(String)
 "j1", true(boolean)

Task a complete to Task b, it works fine; But Task b complete to Task c, occur error.

**Comment by goozi** [comment]
@jbarrez First of all thank you very much for your answer!Problem solved, after the mybatis version update to 3.2.5.
